### ISSUE TITLE
Invalid "possibly unbound variable" warning

### ISSUE THREAD

**Issue by phgmacedo** [microsoft/pyright#3776]
For the following code: 

```python
import pandas as pd

def f(a, b=False, c=False):
    if a.h.any():
        t = a.h.copy()
        a.drop('T', axis=1, inplace=True)
    if c:
        assert a.t.any(), "There's no t!"

    cols = a.columns
    d = a.cumprod(axis=0).copy()
    if not b:
        if c:
            d.insert(len(d.columns), 'T', t.copy(), allow_duplicates=False)
        return d
    s = pd.Series(0, index=list(range(len(d))))
    for i in cols:
        s = s + d[i]
    p = s.to_frame()
    p.columns = list(["P"])
    if c:
        p.insert(len(p.columns), 'T', t.copy(), allow_duplicates=False)
    return p
```
**Describe the bug**
I get a message that `"t" is possibly unbound`. It's guarded by an assertion:

```python
 if c:
        assert a.t.any(), "There's no t!"
```

**Expected behavior**
Code that is guarded by an assertion or exception should not be reported as possdibly unbound. 

**Screenshots or Code**

![dt](https://user-images.githubusercontent.com/92439290/182273563-3883cb3e-e3fc-4fca-9ee3-2394141dc010.png)

**Extra PyClippy crap**:

There's so much shit inserted in the context menu that I almost wasn't able to copy the warning message.

**Comment by erictraut** [comment]
Pyright is working as designed. The symbol `t` is assigned a value within a conditional code path. There are other code paths where it is not assigned a value, so it is potentially unbound.

Type checkers do not track dependencies between variables. Doing so would be prohibitively expensive. A type checker like pyright therefore will not track that the type of `t` is a particular type when `c` is true but a different type when `c` is false. 

My recommendation in cases like this is to initialize the symbol `t` with a value of `None` before the first `if` statement. Later, you can use `t` in a conditional check.

```python
    t = None
    if a.h.any():
        t = a.h.copy()
    ...
    if t:
        p.insert(len(p.columns), "T", t.copy(), allow_duplicates=False)
```
